### PR TITLE
fix: stream memory counting durin snapshot loading

### PR DESCRIPTION
**Pull request by BorysTheDev**
fixes: https://github.com/dragonflydb/dragonfly/issues/4321
problem: we didn't calculate stream memory usage during snapshot loading

### CODE REVIEW

**Comment by adiholden** [comment]
Please add unit test

File: tests/dragonfly/snapshot_test.py [hunk 1]
```diff
@@ -639,3 +639,34 @@ async def check_flag(key, flag):
 
     await check_flag("key1", 2)
     await check_flag("key2", 123456)
+
+
+@pytest.mark.parametrize("format", FILE_FORMATS)
+@dfly_args({**BASIC_ARGS})
+async def test_infomemory_for_stream(df_factory, format: str):
```
Comment (adiholden) [inline tests/dragonfly/snapshot_test.py:646]: why pytest and not unit test?

Comment (adiholden) [reply tests/dragonfly/snapshot_test.py:646]: you can also do dump and restore in the unit test, does not have to be a save

File: tests/dragonfly/snapshot_test.py [hunk 2]
```diff
@@ -639,3 +639,34 @@ async def check_flag(key, flag):
...
+    instance.start()
+    async_client = instance.client()
+    await async_client.execute_command("XADD test 1 var val1")
+    await async_client.execute_command("XADD test 2 var val2")
+
+    async def assert_stream_mem_not_zero():
+        res = await async_client.execute_command("INFO MEMORY")
+
+        for line in res.splitlines():
+            if line.startswith("type_used_memory_stream"):
```
Comment (adiholden) [inline tests/dragonfly/snapshot_test.py:658]: I would check that after adding stream items type_used_memory_stream is > 0, save it to a var and after loading the data again check the value of type_used_memory_stream is the same

File: tests/dragonfly/snapshot_test.py (hunk 1 above)
Comment (adiholden) [reply tests/dragonfly/snapshot_test.py:646]: also we can add checks to test test_rss_used_mem_gap
add a save and reload step there and
do all the memory accounting checks we have there after the reload

File: tests/dragonfly/snapshot_test.py (hunk 2 above)
Comment (BorysTheDev) [reply tests/dragonfly/snapshot_test.py:658]: for simple xadd we do a preallocation so the numbers will be different

Comment (adiholden) [reply tests/dragonfly/snapshot_test.py:658]: ok then lets not check that they are the same value but make sure both > 0

File: src/server/dragonfly_test.cc [hunk 3]
```diff
@@ -802,6 +803,41 @@ TEST_F(DflyEngineTest, DebugObject) {
...
+      stream_mem_first = std::stol(stream_mem[1]);
+    }
+  }
+  EXPECT_GT(stream_mem_first, 0);
+
+  auto dump = Run({"dump", "test"});
+  Run({"del", "test"});
+  Run({"restore", "test", "0", facade::ToSV(dump.GetBuf())});
+
+  resp = Run({"info", "memory"});
```
Comment (adiholden) [inline src/server/dragonfly_test.cc:826]: nit:
you can use GetMetrics().db_stats[0].memory_usage_by_type[OBJ_STREAM]

File: src/server/dragonfly_test.cc [hunk 4]
```diff
@@ -802,6 +803,29 @@ TEST_F(DflyEngineTest, DebugObject) {
   EXPECT_THAT(resp.GetString(), HasSubstr("encoding:listpack"));
 }
 
+TEST_F(DflyEngineTest, StreamMemInfo) {
+  for (int i = 1; i < 2; ++i) {
+    Run({"XADD", "test", std::to_string(i), "var", "val" + std::to_string(i)});
+  }
+
+  auto resp = Run({"info", "memory"});
+  auto str_resp = absl::StrSplit(resp.GetString(), "\r\n");
```
Comment (adiholden) [inline src/server/dragonfly_test.cc:812]: ?

File: src/server/dragonfly_test.cc [hunk 5]
```diff
@@ -10,6 +10,7 @@ extern "C" {
 #include <absl/strings/ascii.h>
 #include <absl/strings/charconv.h>
 #include <absl/strings/str_join.h>
+#include <absl/strings/str_split.h>
```
Comment (adiholden) [inline src/server/dragonfly_test.cc:13]: I believe this is redundant

File: tests/dragonfly/memory_test.py [hunk 6]
```diff
@@ -23,7 +23,10 @@
 # memory it might force the gh runner to run out of memory (since OOM killer might not even
 # get a chance to run).
 @dfly_args({"proactor_threads": 4, "maxmemory": "5gb"})
-async def test_rss_used_mem_gap(df_server: DflyInstance, type, keys, val_size, elements):
+async def test_rss_used_mem_gap(df_factory, type, keys, val_size, elements):
```
Comment (adiholden) [inline tests/dragonfly/memory_test.py:26]: please mark as slow

**Review by adiholden — APPROVED** [review]

**Pull request merged**